flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorassembler/VectorAssembler.java

```diff
@@ -74,64 +85,109 @@ public Table[] transform(Table... inputs) {
         DataStream<Row> output =
                 tEnv.toDataStream(inputs[0])
                         .flatMap(
-                                new AssemblerFunc(getInputCols(), getHandleInvalid()),
+                                new AssemblerFunction(
+                                        getInputCols(), getHandleInvalid(), getInputSizes()),
                                 outputTypeInfo);
         Table outputTable = tEnv.fromDataStream(output);
         return new Table[] {outputTable};
     }
 
-    private static class AssemblerFunc implements FlatMapFunction<Row, Row> {
+    private static class AssemblerFunction implements FlatMapFunction<Row, Row> {
         private final String[] inputCols;
         private final String handleInvalid;
+        private final Integer[] inputSizes;
+        private final boolean keepInvalid;
 
-        public AssemblerFunc(String[] inputCols, String handleInvalid) {
+        public AssemblerFunction(String[] inputCols, String handleInvalid, Integer[] inputSizes) {
             this.inputCols = inputCols;
             this.handleInvalid = handleInvalid;
+            this.inputSizes = inputSizes;
+            keepInvalid = handleInvalid.equals(HasHandleInvalid.KEEP_INVALID);
         }
 
         @Override
         public void flatMap(Row value, Collector<Row> out) {
             int nnz = 0;
             int vectorSize = 0;
             try {
-                for (String inputCol : inputCols) {
-                    Object object = value.getField(inputCol);
-                    Preconditions.checkNotNull(object, "Input column value should not be null.");
-                    if (object instanceof Number) {
-                        nnz += 1;
-                        vectorSize += 1;
-                    } else if (object instanceof SparseVector) {
-                        nnz += ((SparseVector) object).indices.length;
-                        vectorSize += ((SparseVector) object).size();
-                    } else if (object instanceof DenseVector) {
-                        nnz += ((DenseVector) object).size();
-                        vectorSize += ((DenseVector) object).size();
+                for (int i = 0; i < inputCols.length; ++i) {
```

Comment:
This logic is a bit messy here. How about we split it into several functions? e.g.,
`assembleNullValues()` and `assembleNotNullValues()`?